### ISSUE TITLE
The queries behind-the-scene not blocked in the presence of a "behind-the-scene * * block"

### ISSUE THREAD

**Issue by Hifexar** [uBlockOrigin/uBlock-issues#662]
### Prerequisites

- [ ] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [ ] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://old.reddit.com/r/uBlockOrigin/)
    
- [ ] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [ ] uBlock Origin is the only extension
    - [ ] uBlock Origin with default lists/settings
    - [ ] using a new, unmodified browser profile
- [ ] I am running the latest version of uBlock Origin
- [ ] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description

The queries behind-the-scene not blocked in the presence of a "behind-the-scene * * block"

I have already written about this bug in the ticket #651. Now installed uBlock Origin v1.20.3b8, but the bug is still there.

Steps to Reproduce:
1. Create a rule "behind-the-scene * * block".
2. Remove behind-the-scene from the white list.
https://gyazo.com/920613eeb55e052fbd27245ee14b3b16
3. Open several tabs with screenshots on the links from the second and third paragraph. Wait. I don't know how long. Sometimes it works in a minute or two. Sometimes it doesn't work for a long time. Here I do not understand on what principle the hidden request is sent. But after a while we see that such a request is sent and it is not blocked.
https://gyazo.com/570dd984f5884870d23d6b52b6b7806f
4. In umatrix everything works correctly.
https://gyazo.com/9a10d714403f7551ab1081330f2f31cc

Firefox 67.0.4 x64
Windows 7 SP1 x64
uBlock Origin v1.20.3b8

### A specific URL where the issue occurs

[A specific URL is MANDATORY for issue happening on a web page, even if it happens "everywhere"]

### Steps to Reproduce

1. Create a rule "behind-the-scene * * block".
2. Remove behind-the-scene from the white list.
https://gyazo.com/920613eeb55e052fbd27245ee14b3b16
3. Open several tabs with screenshots on the links from the second and third paragraph. Wait. I don't know how long. Here I do not understand on what principle the hidden request is sent. But after a while we see that such a request is sent and it is not blocked.
https://gyazo.com/570dd984f5884870d23d6b52b6b7806f
4. In umatrix everything works correctly.
https://gyazo.com/9a10d714403f7551ab1081330f2f31cc

### Expected behavior:

[What you expected to happen]

### Actual behavior:

[What actually happened]

### Your environment

Firefox 67.0.4 x64
Windows 7 SP1 x64
uBlock Origin v1.20.3b8

* uBlock Origin version:
* Browser Name and version:
* Operating System and version:

**Comment by Hifexar** [comment]
Example of another hidden request that was not blocked. How to reproduce the this request do not know.
https://gyazo.com/7356b63554feb78b0d8d7e792a6b8ce6

**Comment by gorhill** [comment]
The logger is telling you the contexts of the network request, click on it to find more details. If `behind-the-scene` is not the root context, than the rule won't apply on it.

**Comment by Hifexar** [comment]
And there's no way to block it? Only manually for each site?

**Comment by gorhill** [comment]
Please click on the logger entry to find details and share what you see regarding _Partyness_ values.

**Comment by Hifexar** [comment]
I see the context and the parties. But I do not want to work out requests that do not come and some open tab. I can lock manually i.gyazo.com or avatars0.githubusercontent.com adding rule. But tgis no way. 

Maybe for such cases there is a Directive similar to moz-extension-scheme for example?

**Comment by gorhill** [comment]
> Please click on the logger entry to find details and **share what you see regarding Partyness values**.

**Comment by Hifexar** [comment]
https://gyazo.com/484da8463be128a3d69b11b5d30e5582

https://gyazo.com/64f0e3e06ba1e3fa1388b232d1635d0f

**Comment by gorhill** [comment]
I am not able to reproduce what you see. You will have to go through the _Prerequisites_ steps and find a way for me to reproduce.

**Comment by Hifexar** [comment]
All from scratch

1.  Download from hire https://portableapps.com/apps/internet/firefox_portable portable firefox 67.0.4
2. Install uBlock0_1.20.3b8.firefox.signed.xpi
https://gyazo.com/719e5e189a75e51818d413fcc846493d
3. Settings ublock:
https://gyazo.com/cc0201b1b2f3445396f9d5d008caab9b
https://gyazo.com/d4f3b865b1a3a27cc497fa1676259fac
https://gyazo.com/550533c66ae8ac0aea16c0750598dc51
https://gyazo.com/b24fdedc1e80c5fa56c7a684744f9219
https://gyazo.com/6adfd2ac2fdfff239b6555f5b4c00116
4. Open the logger
https://gyazo.com/a0ecccbaa257bf2a67ca433bff7ba0fc
5. From this ticket, open all the links to the screenshots and wait a little.
https://gyazo.com/8e302750b978e1b3be12bb8501c24a43
6. And after a while there appears a lot of unlocked hidden requests.
https://gyazo.com/b0a336883a277faccb15c90f4958e757

This is a completely new installation of firefox. Firefox did not change any settings.

**Comment by gorhill** [comment]
I am unable to reproduce. Are you sure you disclosed everything you did in that new Firefox profile? Those requests almost looks like they might come from an extension.

**Comment by Hifexar** [comment]
"Are you sure you disclosed everything you did in that new Firefox profile? "

The firefox settings do not touch.
Just opened about:config to make sure that the fixed version settings do not affect the portable version. And in the same new version of firefox opened this ticket on github to write a comment. Since the portable version forced to close the current firefox.

I didn't do anything else.

**Comment by gorhill** [comment]
Those tabless requests are to document which URL is something like

    https://i.gyazo.com/thumb/1200/.../...-png.jpg

If you use the _preview_ feature in the logger, how the image looks like?

**Comment by gorhill** [comment]
Finally, got a result: it's from the Firefox thumbnail preview in the home page.

**Comment by gorhill** [comment]
So in conclusion: if you do not want these requests to occur, I believe you just need to disable "Top Sites" in Firefox start page.

**Comment by Hifexar** [comment]
Means ublock such requests do not block? Ublock understands that this is a hidden request. There is no Directive to block such requests, as well as behind-the-scene, the moz-extension-scheme, or others?

**Comment by uBlock-user** [comment]
> Ublock understands that this is a hidden request

uBlock can only block see requests that are sent via webrequest API.

**Comment by gorhill** [comment]
I will have to investigate what happens internally to figure out if anything can be safely done -- a bit difficult since it depends entirely on Firefox firing the network request.

Just remember though that in [this issue](https://github.com/uBlockOrigin/uBlock-issues/issues/651) you wanted the opposite behavior, not having `behind-the-scene` being considered -- which I agreed was an issue.

**Comment by Hifexar** [comment]
"Just remember though that in this issue you wanted the opposite behavior"

No, I wanted the same thing, blocking all hidden requests. And behind-the-scene I myself do not understand how I was then on the white list, it was my mistake or a typo. I found it and fixed it. The ticket I opened because it does not block the moz-extension-scheme. And then I saw that some hidden requests are not blocked too.

**Comment by gorhill** [comment]
Note: I can actually reproduce fairly easily if I use "Delete from history" on the thumbnail -- Firefox will fire a network request to fetch the thumbnail again.

**Comment by gorhill** [comment]
Ok, I consider the behavior by design, as per fix to <https://github.com/uBlockOrigin/uBlock-issues/issues/459>.

First, `doc`  requests are never blocked by rules, so whether a request is tabless or not makes no difference, rules can't block the main document. And even _if_ they were, network request of type `doc` (aka `document` aka `main_frame) are never contextless: the context of these network requests are themselves, as root documents are never embedded into something else.

In conclusion, after investigation, all work as expected.

**Comment by Hifexar** [comment]
In umatrix works here is such:
behind-the-scene * doc block" works.

Is that impossible here?

**Comment by uBlock-user** [comment]
> behind-the-scene * doc block

That's in invalid rule as per uBO --

![](https://i.gyazo.com/ff8f1390253255efa21775dc7cafb8df.png)

**Comment by Hifexar** [comment]
"That's in invalid rule as per uBO"

Never mind. That's not what I asked.

**Comment by Hifexar** [comment]
For such a request, the standard tool creates a URL rule, but it does not work either.
https://gyazo.com/4ce82e35b4588f5bbb16e4b466979049
https://gyazo.com/bc6938983b28ae946dae57d67c61774e

That is, it does not swear at this and allows you to add a non-working rule.

**Comment by Hifexar** [comment]
Still new.

1. Allowed everything in umatrix.
2. After loading the hidden doc, the picture from this doc is loaded. And here is she, too, not is blocked, and this already not doc.
https://gyazo.com/827747f679b632f57b7d32fe56d84858

**Comment by gorhill** [comment]
uMatrix issues have to be reported on uMatrix tracker, uBO is an entirely different extension, there should never be an assumption that uBO should behave like uMatrix or vice versa.

**Comment by Hifexar** [comment]
"uMatrix issues have to be reported on uMatrix tracker"

OK, my last message is not about umatrix, but about uBO.

**Comment by gorhill** [comment]
> The queries of the unlocked hidden query doc is also not blocked

By design, because the context is the parent document, not `behind-the-scene`. The context `behind-the-scene` is assigned if and only if there is no way for uBO to assign a valid context. This is what allowed to remove `behind-the-scene` from default whitelist directives, and thus fix <https://wholeftopenthecookiejar.com/data/extensions/uBlock-Origin/firefox/1.14.18>.

Again, `behind-the-scene` is not the same as tabless, network requests can be tabless and not behind-the-scene because a valid context could be looked up and used.

**Comment by gorhill** [comment]
> Perhaps this is a vestige of some old versions of uBO.

I could spend twice or ten times the amount time I spend on this project, and yet still be unable to guarantee that uBO is perfect and no corner cases exist whatsoever. There is no other explanation for this other than I just didn't take time to write the code to handle that case.

**Comment by Hifexar** [comment]
1. "By design, because the context is the parent document, not behind-the-scene. The context behind-the-scene is assigned if and only if there is no way for uBO to assign a valid context. This is what allowed to remove behind-the-scene from default whitelist directives"

OK, thanks. Became clearer.

2. "Я мог бы потратить в два или десять раз больше времени, которое я трачу на этот проект, и все же не могу гарантировать, что uBO совершенен, и никаких угловых случаев не существует. Нет другого объяснения этому, кроме того, что я просто не потратил время на написание кода для обработки этого случая."

I didn't mean to accuse you of anything. I just reported is not quite adequate behavior of the program. If it's not here, I don't know where else.

3. The fact that the appearance of the Directive "other-tabless-scheme" should not count, I Tode understood. Thanks.

**Comment by gorhill** [comment]
? I just reported is not quite adequate behavior of the program. If it's not here, I don't know where else.

Here is fine, I will prioritize and address according to severity (it's already fixed locally on my side). My point is just that there is no way for me to answer _"Why, then, ublock allows you to do this?"_ when it's an overlooked corner case, no doubt uBO is full of such corner cases in the UI.

**Comment by uBlock-user** [comment]
Just an FYI, these are the only valid request types for Dynamic Filtering in uBO --

```
        '*',
        '3p',
        'image',
        'inline-script',
        '1p-script',
        '3p-script',
        '3p-frame',
```
Anything else will be seen invalid.

**Comment by gorhill** [comment]
These apply to hostname-based rules. There is no valid set for URL-based rules because I want to keep it open-ended so that URL-filtering can be applied to whatever the browser is throwing at the extension. This is why `doc` was not disallowed, so now I have introduce a set of disallowed types for URL-based rules.

**Comment by Hifexar** [comment]
"ust an FYI, these are the only valid request types for Dynamic Filtering in uBO"
"Anything else will be seen invalid."

Still here is such I have works: xmlhttprequest. Right now I'm looking at 1.16.13rc0.

**Comment by Hifexar** [comment]
"his is why doc was not disallowed"

It is a pity that doc can not be blocked, because I really do not want to open at least some doc in addition to what I specify in the address bar of the browser. After all thus it is possible to do imperceptibly from it other quiries.

**Comment by uBlock-user** [comment]
So why not use static filter and block it with `$document` directive ?

**Comment by Hifexar** [comment]
"So why not use static filter and block it with $document directive ?"

1. In the case of tables queries, the such static filter does not work. 
2. Static filter how to block all tabless requests outside existing directives?

**Comment by uBlock-user** [comment]
> In the case of tables queries, the such static filter does not work.
"Works only with a static filter with non tabless queries. The static filter also does not work for this tabless queries."

Nothing to do with requests being tabless or not, behind-the-scene requests are subjected to filtering beginning with 1.15.20+

**Comment by Hifexar** [comment]
"Nothing to do with requests being tabless or not, behind-the-scene requests are subjected to filtering beginning with 1.15.20+"

As gorhill explained above, behind-the-scene are queries that have no context defined. But personally for me there is no difference, the context is defined or not, I do not want any of these requests. Only if they are really needed in some site I need for its performance.

**Comment by uBlock-user** [comment]
> I do not want any of these requests

Then you're better off using uMatrix for blocking ALL behind-the-scene + tabless requests.

**Comment by Hifexar** [comment]
"Then you're better off using uMatrix for blocking ALL behind-the-scene + tabless requests."

That's what I do. There all these requests are considered to be behind-the-scene.
Just at first it was strange that ublock misses this. 
Ideally, of course, it would add unblock functions to umatrix post-filtering. That would be great.
